### PR TITLE
Add PKCE verification support

### PR DESCRIPTION
**Pull request by skycocker**
This is pretty much a 1:1 copy of https://github.com/omniauth/omniauth_openid_connect/pull/80/commits with addressed PR suggestions and removed unused code.

You may want to review it with `?w=1` appended to the URL - it'll make reading much easier.

Also, there's one small improvement to the original PR: https://github.com/omniauth/omniauth_openid_connect/pull/128/files#diff-ee829bbbbd9d35ee93126d96347fffbfa8756868001c1d4ad6aff9c4650e8151R251

This allows providing a parameter in case one does not use sessions.

### CODE REVIEW

File: lib/omniauth/strategies/openid_connect.rb
Comment (skycocker) [inline lib/omniauth/strategies/openid_connect.rb:64]: original discussion: https://github.com/omniauth/omniauth_openid_connect/pull/80/files#r802033038

This hasn't been addressed, because these digests are not interchangeable and the feature does not work with just `Digest::SHA2.hexdigest(verifier)`.

File: lib/omniauth/strategies/openid_connect.rb
Comment (skycocker) [inline lib/omniauth/strategies/openid_connect.rb:64]: original discussion: https://github.com/omniauth/omniauth_openid_connect/pull/80/files#r802033038

This hasn't been addressed, because these digests are not interchangeable and the feature does not work with just Digest::SHA2.hexdigest(verifier).

Comment (stanhu) [reply lib/omniauth/strategies/openid_connect.rb:64]: Thanks for checking that.

**Review by stanhu — APPROVED** [review]
Looks good to me. Thanks!

**Comment by stanhu** [comment]
@omniauth/omniauth-oidc Can anyone else review these changes?

**Comment by danjay** [comment]
lgtm too. I'll close my PR now the improvements have been implemented in this one.

**Comment by skycocker** [comment]
@BobbyMcWho @pboling ping ping


this should introduce no changes to people that already have been using https://github.com/omniauth/omniauth_openid_connect/pull/80, except they would now be able to use a version from the master upstream

**Comment by BobbyMcWho** [comment]
I unfortunately don't know much about this strategy, I'll have to find some time to get up to speed on what this proposes

**Comment by skycocker** [comment]
> I unfortunately don't know much about this strategy, I'll have to find some time to get up to speed on what this proposes

here's a good, simple explanation: https://doorkeeper.gitbook.io/guides/ruby-on-rails/pkce-flow

here's less simplified, but still a good one: https://developer.okta.com/docs/guides/implement-grant-type/authcodepkce/main/#about-the-authorization-code-grant-with-pkce

Not supporting PKCE these days excludes you from a lot of OIDC providers.

**Review by pboling — APPROVED** [review]
LGTM

**Review by BobbyMcWho — APPROVED** [review]

**Comment by BobbyMcWho** [comment]
I'm going to merge this, but I don't manage releases for this gem, @stanhu looks like the last one who pushed this to rubygems, so I'll defer to them to release

**Comment by skycocker** [comment]
Thank youuu! Before the release, we may want to merge https://github.com/omniauth/omniauth_openid_connect/pull/130 and https://github.com/omniauth/omniauth_openid_connect/pull/127, so the complete support of the PKCE flow with optional checks is added as a feature.

**Comment by stanhu** [comment]
v0.5.0 has been released.

**Pull request merged**
